### ISSUE TITLE
yum package install failure `yum.Errors.YumBaseError: Error: rpmdb open failed`

### ISSUE THREAD

**Issue by scrummyin** [chef/chef#11266]
## Description
  There seems to be a case where `yum_helper.py` can leave the yum db in a corrupted state.  We have noticed this when using chef to install yum packages via the packer chef-solo. I saw the note at the top of the bug about `chef-solo` plugin bugs not belonging here, but the issue tracker is for code in this repo and `lib/chef/provider/package/yum/yum_helper.py` lives here. After patching that file, I no longer experience the bug. Which makes me things the bug report belongs here.
  
  We are attempting to run a fairly simple yum install of `unzip`. The reason I suspect `yum_helper.py` is that I created a patched rpm with an updated version of that file and have not yet been able to reproduce the error (PR will be attached containing the code).

  I realize that the version I produced may not meet all your requirements (python 2.4), but I wanted to propose a fix for further discussion. As I would be willing to help with creating a python 2.4 version of the file, but I have the hope that, since now even python 2.7 is [being sunsettted](https://www.python.org/doc/sunset-python-2/), python 2.4 support might not be needed any longer.

## Chef Version
`16.11.7`

## Platform Version
centos 7

## Replication Case
This is a "fun" one because we can only get this to repeat ~50% of the time on chef `16.11.7`, by that I mean I ran this 4x earlier today in order to replicate and 2 passed and 2 failed.

1. create a cookbook containing the following

```ruby
yum_package 'unzip' do
  action :install
 end 
 ```
 
 2. apply cookbook to a recently booted centos 7 image.

## Client Output
The console output looks like below:
```
Error executing action `install` on resource 'yum_package[unzip]'
  ================================================================================
                                                                                                                  
  RuntimeError
  ------------           
  yum-helper.py had stderr/stdout output:
                     
  error: rpmdb: BDB0113 Thread/process 2290/140390376314944 failed: BDB1507 Thread died in Berkeley DB library
  error: db5 error(-30973) from dbenv->failchk: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery
  error: cannot open Packages index using db5 -  (-30973)
  error: cannot open Packages database in /var/lib/rpm
  error: rpmdb: BDB0113 Thread/process 2290/140390376314944 failed: BDB1507 Thread died in Berkeley DB library
  error: db5 error(-30973) from dbenv->failchk: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery
  error: cannot open Packages database in /var/lib/rpm
  Traceback (most recent call last):
    File "/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.11.7/lib/chef/provider/package/yum/yum_helper.py", line 209, in <module>
      query(command)
    File "/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.11.7/lib/chef/provider/package/yum/yum_helper.py", line 88, in query
      enabled_repos = base.repos.listEnabled()                       
    File "/usr/lib/python2.7/site-packages/yum/__init__.py", line 1071, in <lambda>
      repos = property(fget=lambda self: self._getRepos(),
    File "/usr/lib/python2.7/site-packages/yum/__init__.py", line 686, in _getRepos
      self.conf # touch the config class first
    File "/usr/lib/python2.7/site-packages/yum/__init__.py", line 1079, in <lambda>
      conf = property(fget=lambda self: self._getConfig(),
    File "/usr/lib/python2.7/site-packages/yum/__init__.py", line 349, in _getConfig
      startupconf = config.readStartupConfig(fn, root, releasever)
    File "/usr/lib/python2.7/site-packages/yum/config.py", line 1115, in readStartupConfig    
      startupconf.distroverpkg)                                                         
    File "/usr/lib/python2.7/site-packages/yum/config.py", line 1260, in _getsysver
      raise Errors.YumBaseError("Error: " + str(e))                                                                                    
  yum.Errors.YumBaseError: Error: rpmdb open failed
                                                                                                              
  Resource Declaration:                                                                                   
  ---------------------                                  
  # In /tmp/packer-chef-solo/local-mode-cache/cache/cookbooks/mybooks/recipes/default.rb
                                                                                                              
    9: yum_package 'unzip' do                                                                             
   10:   action :install                              
   11: end                          
   12:                                                                                                                                
                    
  Compiled Resource:                                                                                                              
  ------------------                          
  # Declared in /tmp/packer-chef-solo/local-mode-cache/cache/cookbooks/mybooks/recipes/default.rb:9:in `from_file'
                                                          
  yum_package("unzip") do                                                          
    package_name "unzip"                      
    action [:install]                                                              
    default_guard_interpreter :default                    
    declared_type :yum_package                                                      
    cookbook_name "mybooks"                                       
    recipe_name "default"                                                                 
  end
  
  System Info:                                     
  ------------                                     
  chef_version=16.11.7                                     
  platform=centos
  platform_version=7.9.2009
  ruby=ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux]
  program_name=/bin/chef-solo
  executable=/opt/chef/bin/chef-sol

Running handlers:
  ERROR: Running exception handlers
Running handlers complete
  ERROR: Exception handlers complete
Chef Infra Client failed. 0 resources updated in 39 seconds
  FATAL: Stacktrace dumped to /tmp/packer-chef-solo/local-mode-cache/cache/chef-stacktrace.out
  FATAL: Please provide the contents of the stacktrace.out file if you file a bug report

  FATAL: RuntimeError: yum_package[unzip] (mybooks::default line 9) had an error: RuntimeError: yum-helper.py had stderr/stdout output:

  error: rpmdb: BDB0113 Thread/process 2290/140390376314944 failed: BDB1507 Thread died in Berkeley DB library
  error: db5 error(-30973) from dbenv->failchk: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery
  error: cannot open Packages index using db5 -  (-30973)
  error: cannot open Packages database in /var/lib/rpm
  error: rpmdb: BDB0113 Thread/process 2290/140390376314944 failed: BDB1507 Thread died in Berkeley DB library
  error: db5 error(-30973) from dbenv->failchk: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery
  error: cannot open Packages database in /var/lib/rpm
  Traceback (most recent call last):
    File "/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.11.7/lib/chef/provider/package/yum/yum_helper.py", line 209, in <module>
      query(command)
    File "/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.11.7/lib/chef/provider/package/yum/yum_helper.py", line 88, in query
      enabled_repos = base.repos.listEnabled()
    File "/usr/lib/python2.7/site-packages/yum/__init__.py", line 1071, in <lambda>
      repos = property(fget=lambda self: self._getRepos(),
    File "/usr/lib/python2.7/site-packages/yum/__init__.py", line 686, in _getRepos
      self.conf # touch the config class first
    File "/usr/lib/python2.7/site-packages/yum/__init__.py", line 1079, in <lambda>
      conf = property(fget=lambda self: self._getConfig(),
    File "/usr/lib/python2.7/site-packages/yum/__init__.py", line 349, in _getConfig
      startupconf = config.readStartupConfig(fn, root, releasever)
    File "/usr/lib/python2.7/site-packages/yum/config.py", line 1115, in readStartupConfig
      startupconf.distroverpkg)
    File "/usr/lib/python2.7/site-packages/yum/config.py", line 1260, in _getsysver
      raise Errors.YumBaseError("Error: " + str(e))
  yum.Errors.YumBaseError: Error: rpmdb open failed
```
From the above I believe the most important line is `FATAL: RuntimeError: yum_package[unzip] (mybooks::default line 9) had an error: RuntimeError: yum-helper.py had stderr/stdout output:` which is interesting because this recent commit reads to me as [this could happen but has not been seen](https://github.com/chef/chef/commit/dccabff4042ca8d556f6f56f6ae87da6265150bf#diff-f63177662974b0d0626474f6759886c0bc915ae1cfcde0d221fc75a9c358167c).

## Stacktrace
https://gist.github.com/scrummyin/717388a2f4f74b97d1285bb564c98fe2#file-chef-stacktrace-out

**Comment by tas50** [comment]
This is potentially fixed with the upcoming 17.1 release

**Comment by lamont-granquist** [comment]
I think this one got fixed
